# Bigramas

A análise de bigramas básicamente serve para visualizar quais palavras aparecem frequentemente aparecem junta, sendo assim revelando combinações comuns de palavras (ou palavras compostas) e fornece insights sobre a estrutura do texto e o contexto das palavras

Essa analise vai ser feitas com todo os site que nossa equipe selecionou que pode ter a ver com o nosso tema, o csv utilizado é 'dados_sites.csv'.

In [20]:
import csv
import nltk
from nltk import word_tokenize
from nltk.util import ngrams
from nltk.corpus import stopwords
from collections import Counter
import string
import pandas as pd

# Definir as palavras indesejadas
palavras_indesejadas = set(stopwords.words('portuguese'))

# Definir o limite de frequência mínima
limite_frequencia = 2

# Lê o arquivo CSV com a codificação correta
with open('dados_sites.csv', 'r', encoding='utf-8') as arquivo_csv:
    leitor_csv = csv.reader(arquivo_csv)
    
    # Pula o cabeçalho do CSV
    next(leitor_csv)
    
    # Lista para armazenar os bigramas relevantes
    bigramas_relevantes = []
    
    # Itera sobre as linhas do CSV
    for linha in leitor_csv:
        texto = linha[1]  # Obtém o texto da segunda coluna
        
        # Tokeniza o texto em palavras
        tokens = word_tokenize(texto)
        
        # Filtra as palavras indesejadas, remove pontuações e números
        palavras_filtradas = [palavra.lower() for palavra in tokens if palavra.lower() not in palavras_indesejadas and palavra not in string.punctuation and not any(char.isdigit() for char in palavra)]
        
        # Gera os bigramas
        bigramas = list(ngrams(palavras_filtradas, 2))
        
        # Atualiza a lista de bigramas relevantes com base no limite de frequência
        bigramas_relevantes.extend([bigrama for bigrama in bigramas if bigramas.count(bigrama) >= limite_frequencia])

# Conta a frequência dos bigramas relevantes
frequencia_bigramas = Counter(bigramas_relevantes)

# Cria o dataframe com os bigramas e suas frequências
df = pd.DataFrame.from_dict(frequencia_bigramas, orient='index', columns=['Frequência'])
df.index.names = ['Bigramas']
df.sort_values(by='Frequência', ascending=False, inplace=True)

# Configura para exibir todas as linhas do dataframe
pd.set_option('display.max_rows', None)


Lembrando que as stopwords foram retiradas, ou seja bigramas como (fim,semana) aparecem em conjunto pois antes eram "fim de semana"

In [17]:
df = df[df['Frequência'] > 30]
df

Frequência
Bigramas                                      
(exercícios, sobre)                       5294
(festa, junina)                           4332
(dúvidas, sobre)                          2950
(..., exercícios)                         2843
(guia, folha)                             2296
(exclusivo, assinantes)                   2009
(fim, semana)                             2004
(faça, login)                             1975
(salvar, artigos)                         1973
(assine, faça)                            1973
(recurso, exclusivo)                      1973
(artigos, recurso)                        1973
(assinantes, assine)                      1973
(ler, salvar)                             1938
(salvar, ler)                             1938
(virada, cultural)                        1733
(bares, noite)                            1728
(semana, sp)                              1591
(clique, aqui)                            1520
(sp, shows)                               1501
(sp, restaurantes)                        1469
(sp, passeios)                            1408
(teste, conhecimentos)                    1252
(saiba, comprar)                          1250
(show, sp)                                1249
(sp, veja)                                1240
(months, cookie)                          1205
(cookie, definido)                        1200
(folha, s.paulo)                          1175
(ingressos, shows)                        1122
(rita, lee)                               1090
(dia, mães)                               1075
(casas, shows)                            1073
(sp, melhor)                              1065
(moda, beleza)                            1024
(cookie, consent)                         1006
(categoria, ``)                           1005
(cookies, categoria)                      1005
(consentimento, usuário)                  1005
(usuário, cookies)                        1005
(gdpr, cookie)                            1000
(plug-in, gdpr)                           1000
(definido, plug-in)                       1000
(sp, saiba)                                973
(saiba, sobre)                             942
(restaurantes, bares)                      929
(fonte, instagram)                         924
(veja, programação)                        919
(grande, conquista)                        913
(cinema, streaming)                        859
(melhor, casa)                             852
(casa, show)                               850
(teatro, shows)                            842
(streaming, restaurantes)                  840
(shows, passeios)                          840
(folha, guia)                              840
(noite, teatro)                            840
(logo, guia)                               840
(comprar, ingressos)                       836
(armazenar, consentimento)                 804
(usado, armazenar)                         804
(your, browser)                            802
(restaurantes, conheça)                    801
(famosos, novelas)                         800
(·, festas)                                800
(beleza, saúde)                            800
(saiba, quais)                             777
(brasil, escola)                           765
(conhecimentos, sobre)                     750
(you, want)                                745
(hours, hours)                             735
(festas, juninas)                          733
(melhor, fim)                              709
(reality, shows)                           704
(passo, passo)                             703
(fala, brasil)                             699
(the, best)                                685
(mostra, cinema)                           684
(passeios, exposição)                      683
(want, to)                                 676
(política, privacidade)                    675
(dia, trabalho)                            665
(fashion, bubbles)                         651
(história, brasil)                       

In [18]:
len(df)

1671

# Insights

#### bigramas referente ao público-alvo

1. (brasil,escola) com frequencia de 765

2. (passeios, crianças) com frequencia de 560

3. (ensino,médio) com frequencia de 317

4. (diversão, educando) com frequencia de 256

5. (crianças, bebês) com frequencia de 204

6. (junina, crianças) com frequencia de 198

7. (bebês, fotos) com frequencia de 198

8. (ensino, fundamental) com frequencia de 83


*Esses bigramas podem estar indicando a necessidade do nosso grupo investir em um publico-alvo mais jovem, como crianças e adolescentes visto que eles participam desses eventos frequentemente nas escolas.*

#### bigramas referentes a época

1. (fim, semana) com frequencia de 2004

2. (final,semana) com frequencia de 349

*Esses bigramas podem indicar qual época nosso clientes buscam mais nosso produtos, podemos validar esse teória vendo a frequencia da citações sobre o tema nas redes sociais ao decorrer da semana.*

#### bigramas referentes a local

1. (parque, ibirapuera) com frequencia de 266

*Talvez o parque ibirapuera tenha um relação com nosso público, podemos analisar a existencia de eventos e shows nessa época do ano*